File: python/avro_example.py
import avro.schema
from avro.io import DatumWriter, DatumReader
from io import BytesIO

schema_str = """
{
  "type": "record",
  "name": "User",
  "fields": [
    { "name": "name", "type": "string" },
    { "name": "id", "type": "int" }
  ]
}
"""
schema = avro.schema.parse(schema_str)

user = {
    "name": "Matteo Antony",
    "id": 123
}

writer = DatumWriter(schema)
bytes_writer = BytesIO()
encoder = avro.io.BinaryEncoder(bytes_writer)
writer.write(user, encoder)
avro_bytes = bytes_writer.getvalue()
print("Avro serialized (bytes): " + str(len(avro_bytes)) + " bytes")

reader = DatumReader(schema)
bytes_reader = BytesIO(avro_bytes)
decoder = avro.io.BinaryDecoder(bytes_reader)
deserialized_user = reader.read(decoder)
print("Deserialized dictionary:")
print(deserialized_user)
print("Name: " + deserialized_user["name"])

File: it/java/src/main/java/AvroExample.java
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;

import org.apache.avro.Schema;
import org.apache.avro.generic.GenericData;
import org.apache.avro.generic.GenericRecord;
import org.apache.avro.io.DatumReader;
import org.apache.avro.io.DatumWriter;
import org.apache.avro.io.Decoder;
import org.apache.avro.io.DecoderFactory;
import org.apache.avro.io.Encoder;
import org.apache.avro.io.EncoderFactory;
import org.apache.avro.specific.SpecificDatumReader;
import org.apache.avro.specific.SpecificDatumWriter;

public class AvroExample {
  public static void main(String[] args) throws IOException {
    // 1. Schema definition
    String schemaStr = "{"
        + "\"type\": \"record\","
        + "\"name\": \"User\","
        + "\"fields\": ["
        + "{ \"name\": \"name\", \"type\": \"string\" },"
        + "{ \"name\": \"id\", \"type\": \"int\" }"
        + "]"
        + "}";

    Schema schema = new Schema.Parser().parse(schemaStr);

    GenericRecord user = new GenericData.Record(schema);
    user.put("name", "Matteo Antony");
    user.put("id", 123);

    DatumWriter<GenericRecord> datumWriter = new SpecificDatumWriter<>(schema);
    ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
    Encoder encoder = EncoderFactory.get().binaryEncoder(outputStream, null);
    datumWriter.write(user, encoder);
    encoder.flush();
    byte[] avroBytes = outputStream.toByteArray();
    outputStream.close();

    System.out.println("Avro Serialized (bytes): " + avroBytes.length + " bytes");

    DatumReader<GenericRecord> datumReader = new SpecificDatumReader<>(schema);
    ByteArrayInputStream inputStream = new ByteArrayInputStream(avroBytes);
    Decoder decoder = DecoderFactory.get().binaryDecoder(inputStream, null);
    GenericRecord deserializedUser = datumReader.read(null, decoder);
    inputStream.close();

    System.out.println("Deserialized name: " + deserializedUser.get("name"));
  }
}
